source/agora/consensus/Reward.d

```diff
@@ -0,0 +1,413 @@
+    {
+        // Calculate payout number this one will be
+        auto payouts = height / this.payout_period;
+
+        // Sum up all the rewards so far
+        Amount rewards = this.commonsPerPayout; rewards.mul(payouts);
+
+        // If the allocation is all paid out
+        try
+        {
+            ensure(rewards < this.totalCommonsBudgetRewards,
+                "No more rewards for Commons Budget at height {}", height);
+        }
+        catch (Exception e)
+        {
+            log.trace(e.msg);
+            return 0.coins;
+        }
+
+        return this.commonsPerPayout;
+    }
+
+    /***************************************************************************
+
+        Returns the total allocated whitepaper validator reward.
+
+        Params:
+            height = last block of the current payout period
+
+        Returns:
+            validator reward
+
+    ***************************************************************************/
+
+    private Amount allocatedValRewards (in Height height) nothrow @safe
+    {
+        // we use previous block height to only reduce after last payment of year
+        ulong payoutYear = (this.block_interval_sec * (height - 1)) / yearOfSecs;
+
+        // Initially set to first year in coin units and doing the maths using ulong
+        ulong reward_value = firstYearValRewards;
+
+        // Apply the yearly reduction of 6.31% of rewards to Validators
+        iota(payoutYear).each!((ulong y)
+        {
+            // Reduction of 6.31%
+            //  == multiply by 93.69%
+            //  == multiply by 9369 then divide by 1000
+            // This is done to minimize losses when calculating as we are using ulong not double
+            reward_value = (reward_value * 9_369) / 10_000;
+
+            log.trace("allocatedValRewards(height={}): Reduced by 6.31%. Yearly is {} for year {}",
+                height, reward_value, y + 2);
+        });
+        // Now divide this year's reward by number of payouts
+        reward_value = reward_value / this.payouts_per_year;
+
+        Amount reward = Amount(reward_value);
+        log.trace("allocatedValRewards(height={}): reward is {}", height, reward_value);
+
+        return reward;
+    }
+
+    /***************************************************************************
+
+        Returns the total validator reward after applying the penalty for
+        missing signatures.
+
+        Params:
+            total_allocated = allocated `Amount` to pay the Validators if all
+                blocks have all signatures
+            percent_signed = (#actual signatures / #expected signatures) * 100
+                on all the blocks for payout period
+
+        Returns:
+            validator reward - penalty on validators due to missing signature
+
+    ***************************************************************************/
+
+    private Amount reducedValRewards (in Amount total_allocated, ubyte percent_signed) nothrow @safe
```

Comment:
This is `private` but never called from another function ?

Reply:
It is called in `calculateBlockRewards`.